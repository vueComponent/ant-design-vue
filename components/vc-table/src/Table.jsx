/* eslint-disable camelcase */
import shallowequal from 'shallowequal';
import merge from 'lodash/merge';
import classes from 'component-classes';
import classNames from 'classnames';
import PropTypes from '../../_util/vue-types';
import { debounce, getColumnsKey, validateValue } from './utils';
import warning from '../../_util/warning';
import addEventListener from '../../vc-util/Dom/addEventListener';
import ColumnManager from './ColumnManager';
import HeadTable from './HeadTable';
import BodyTable from './BodyTable';
import ExpandableTable from './ExpandableTable';
import { initDefaultProps, getOptionProps, getListeners } from '../../_util/props-util';
import BaseMixin from '../../_util/BaseMixin';
import Vue from 'vue';
import StickyOffsets from './useStickyOffsets';
import { getCellFixedInfo } from './fixUtil';
import ResizeObserver from '../../vc-resize-observer';

export default {
  name: 'Table',
  mixins: [BaseMixin, StickyOffsets],
  provide() {
    return { 'table-store': this.store, table: this };
  },
  props: initDefaultProps(
    {
      data: PropTypes.array,
      useFixedHeader: PropTypes.bool,
      columns: PropTypes.array,
      prefixCls: PropTypes.string,
      bodyStyle: PropTypes.object,
      rowKey: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
      rowClassName: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
      customRow: PropTypes.func,
      customHeaderRow: PropTypes.func,
      // onRowClick: PropTypes.func,
      // onRowDoubleClick: PropTypes.func,
      // onRowContextMenu: PropTypes.func,
      // onRowMouseEnter: PropTypes.func,
      // onRowMouseLeave: PropTypes.func,
      showHeader: PropTypes.bool,
      title: PropTypes.func,
      id: PropTypes.string,
      footer: PropTypes.func,
      emptyText: PropTypes.any,
      scroll: PropTypes.object,
      rowRef: PropTypes.func,
      getBodyWrapper: PropTypes.func,
      components: PropTypes.shape({
        table: PropTypes.any,
        header: PropTypes.shape({
          wrapper: PropTypes.any,
          row: PropTypes.any,
          cell: PropTypes.any,
        }),
        body: PropTypes.shape({
          wrapper: PropTypes.any,
          row: PropTypes.any,
          cell: PropTypes.any,
        }),
      }),
      expandIconAsCell: PropTypes.bool,
      expandedRowKeys: PropTypes.array,
      expandedRowClassName: PropTypes.func,
      defaultExpandAllRows: PropTypes.bool,
      defaultExpandedRowKeys: PropTypes.array,
      expandIconColumnIndex: PropTypes.number,
      expandedRowRender: PropTypes.func,
      childrenColumnName: PropTypes.string,
      indentSize: PropTypes.number,
      expandRowByClick: PropTypes.bool,
      expandIcon: PropTypes.func,
      tableLayout: PropTypes.string,
      transformCellText: PropTypes.func,
    },
    {
      data: [],
      useFixedHeader: false,
      rowKey: 'key',
      rowClassName: () => '',
      prefixCls: 'rc-table',
      bodyStyle: {},
      showHeader: true,
      scroll: {},
      rowRef: () => null,
      emptyText: () => 'No Data',
      customHeaderRow: () => {},
    },
  ),
  data() {
    this.preData = [...this.data];
    this.store = (this.$root.constructor.observable || Vue.observable)({
      currentHoverKey: null,
      fixedColumnsHeadRowsHeight: [],
      fixedColumnsBodyRowsHeight: {},
      expandedRowsHeight: {},
      expandedRowKeys: [],
      fixedInfoList:[],
      stickyOffsets:{},
    });

    return {
      columnManager: new ColumnManager(this.columns),
      fixedInfoList: [],
      pingedLeft: false,
      pingedRight: false,
      sComponents: merge(
        {
          table: 'table',
          header: {
            wrapper: 'thead',
            row: 'tr',
            cell: 'th',
          },
          body: {
            wrapper: 'tbody',
            row: 'tr',
            cell: 'td',
          },
        },
        this.components,
      ),
    };
  },
  computed:{
    colsKeys(){
      return getColumnsKey(this.columnManager.leafColumns());
    },
    horizonScroll({scroll}){
      return scroll && validateValue(scroll.x);
    },
  },
  watch: {
    components() {
      this._components = merge(
        {
          table: 'table',
          header: {
            wrapper: 'thead',
            row: 'tr',
            cell: 'th',
          },
          body: {
            wrapper: 'tbody',
            row: 'tr',
            cell: 'td',
          },
        },
        this.components,
      );
    },
    columns(val) {
      if (val) {
        this.columnManager.reset(val);
      }
    },
    data(val) {
      if (val.length === 0 && this.hasScrollX()) {
        this.$nextTick(() => {
          this.resetScrollX();
        });
      }
    },
    colsWidths:{
      immediate: true,
      handler(){
        const pureColWidths = this.colsKeys.map(columnKey => this.colsWidths.get(columnKey));
        this.colWidths = pureColWidths;
        this.stickyColumns = this.columnManager.leafColumns();
      },
    },
    stickyOffsets:{
      immediate:true,
      handler(val){
        this.store.stickyOffsets = val;
        this.store.fixedInfoList = this.fixedInfoList = this.columnManager.leafColumns().map((_, colIndex) =>
          getCellFixedInfo(colIndex, colIndex, this.columnManager.leafColumns(), val),
        );
      },
    },
  },

  // static childContextTypes = {
  //   table: PropTypes.any,
  //   components: PropTypes.any,
  // },

  created() {
    ['rowClick', 'rowDoubleclick', 'rowContextmenu', 'rowMouseenter', 'rowMouseleave'].forEach(
      name => {
        warning(
          getListeners(this)[name] === undefined,
          `${name} is deprecated, please use customRow instead.`,
        );
      },
    );

    warning(
      this.getBodyWrapper === undefined,
      'getBodyWrapper is deprecated, please use custom components instead.',
    );

    this.setScrollPosition('left');

    this.debouncedWindowResize = debounce(this.handleWindowResize, 150);
  },

  mounted() {
    this.$nextTick(() => {
      if (this.columnManager.isAnyColumnsFixed()) {
        this.handleWindowResize();
        this.resizeEvent = addEventListener(window, 'resize', this.debouncedWindowResize);
      }
      // https://github.com/ant-design/ant-design/issues/11635
      if (this.ref_headTable) {
        this.ref_headTable.scrollLeft = 0;
      }
      if (this.ref_bodyTable) {
        this.ref_bodyTable.scrollLeft = 0;
      }
      this.onFullTableResize();
    });
  },

  updated() {
    this.$nextTick(() => {
      if (this.columnManager.isAnyColumnsFixed()) {
        this.handleWindowResize();
        if (!this.resizeEvent) {
          this.resizeEvent = addEventListener(window, 'resize', this.debouncedWindowResize);
        }
      }
      this.onFullTableResize();
    });
  },

  beforeDestroy() {
    if (this.resizeEvent) {
      this.resizeEvent.remove();
    }
    if (this.debouncedWindowResize) {
      this.debouncedWindowResize.cancel();
    }
  },
  methods: {
    getRowKey(record, index) {
      const rowKey = this.rowKey;
      const key = typeof rowKey === 'function' ? rowKey(record, index) : record[rowKey];
      warning(
        key !== undefined,
        'Each record in table should have a unique `key` prop,' +
          'or set `rowKey` to an unique primary key.',
      );
      return key === undefined ? index : key;
    },

    setScrollPosition(position) {
      this.scrollPosition = position;
      if (this.tableNode) {
        const { prefixCls } = this;
        if (position === 'both') {
          classes(this.tableNode)
            .remove(new RegExp(`^${prefixCls}-scroll-position-.+$`))
            .add(`${prefixCls}-scroll-position-left`)
            .add(`${prefixCls}-scroll-position-right`);
        } else {
          classes(this.tableNode)
            .remove(new RegExp(`^${prefixCls}-scroll-position-.+$`))
            .add(`${prefixCls}-scroll-position-${position}`);
        }
      }
    },

    setScrollPositionClassName() {
      const node = this.ref_bodyTable;
      const scrollToLeft = node.scrollLeft === 0;
      const scrollToRight =
        node.scrollLeft + 1 >=
        node.children[0].getBoundingClientRect().width - node.getBoundingClientRect().width;
      if (scrollToLeft && scrollToRight) {
        this.setScrollPosition('both');
      } else if (scrollToLeft) {
        this.setScrollPosition('left');
      } else if (scrollToRight) {
        this.setScrollPosition('right');
      } else if (this.scrollPosition !== 'middle') {
        this.setScrollPosition('middle');
      }
    },

    isTableLayoutFixed() {
      const { tableLayout, columns = [], useFixedHeader, scroll = {} } = this.$props;
      if (typeof tableLayout !== 'undefined') {
        return tableLayout === 'fixed';
      }
      // if one column is ellipsis, use fixed table layout to fix align issue
      if (columns.some(({ ellipsis }) => !!ellipsis)) {
        return true;
      }
      // if header fixed, use fixed table layout to fix align issue
      if (useFixedHeader || scroll.y) {
        return true;
      }
      // if scroll.x is number/px/% width value, we should fixed table layout
      // to avoid long word layout broken issue
      if (scroll.x && scroll.x !== true && scroll.x !== 'max-content') {
        return true;
      }
      return false;
    },

    handleWindowResize() {
      this.syncFixedTableRowHeight();
      this.setScrollPositionClassName();
    },

    syncFixedTableRowHeight() {
      const tableRect = this.tableNode.getBoundingClientRect();
      // If tableNode's height less than 0, suppose it is hidden and don't recalculate rowHeight.
      // see: https://github.com/ant-design/ant-design/issues/4836
      if (tableRect.height !== undefined && tableRect.height <= 0) {
        return;
      }
      const { prefixCls } = this;
      const headRows = this.ref_headTable
        ? this.ref_headTable.querySelectorAll('thead')
        : this.ref_bodyTable.querySelectorAll('thead');
      const bodyRows = this.ref_bodyTable.querySelectorAll(`.${prefixCls}-row`) || [];
      const fixedColumnsHeadRowsHeight = [].map.call(headRows, row =>
        row.getBoundingClientRect().height ? row.getBoundingClientRect().height - 0.5 : 'auto',
      );
      const state = this.store;
      const fixedColumnsBodyRowsHeight = [].reduce.call(
        bodyRows,
        (acc, row) => {
          const rowKey = row.getAttribute('data-row-key');
          const height =
            row.getBoundingClientRect().height ||
            state.fixedColumnsBodyRowsHeight[rowKey] ||
            'auto';
          acc[rowKey] = height;
          return acc;
        },
        {},
      );
      if (
        shallowequal(state.fixedColumnsHeadRowsHeight, fixedColumnsHeadRowsHeight) &&
        shallowequal(state.fixedColumnsBodyRowsHeight, fixedColumnsBodyRowsHeight)
      ) {
        return;
      }
      this.store.fixedColumnsHeadRowsHeight = fixedColumnsHeadRowsHeight;
      this.store.fixedColumnsBodyRowsHeight = fixedColumnsBodyRowsHeight;
    },

    resetScrollX() {
      if (this.ref_headTable) {
        this.ref_headTable.scrollLeft = 0;
      }
      if (this.ref_bodyTable) {
        this.ref_bodyTable.scrollLeft = 0;
      }
    },

    hasScrollX() {
      const { scroll = {} } = this;
      return 'x' in scroll;
    },

    handleBodyScrollLeft(e) {
      // Fix https://github.com/ant-design/ant-design/issues/7635
      if (e.currentTarget !== e.target) {
        return;
      }
      const target = e.target;
      const { scroll = {} } = this;
      const { ref_headTable, ref_bodyTable } = this;
      if (target.scrollLeft !== this.lastScrollLeft && scroll.x) {
        if (target === ref_bodyTable && ref_headTable) {
          ref_headTable.scrollLeft = target.scrollLeft;
        } else if (target === ref_headTable && ref_bodyTable) {
          ref_bodyTable.scrollLeft = target.scrollLeft;
        }
        this.setScrollPositionClassName();
      }
      // Remember last scrollLeft for scroll direction detecting.
      this.lastScrollLeft = target.scrollLeft;
    },

    handleBodyScrollTop(e) {
      const target = e.target;
      // Fix https://github.com/ant-design/ant-design/issues/9033
      if (e.currentTarget !== target) {
        return;
      }
      const { scroll = {} } = this;
      const {
        ref_headTable,
        ref_bodyTable,
        ref_fixedColumnsBodyLeft,
        ref_fixedColumnsBodyRight,
      } = this;
      if (target.scrollTop !== this.lastScrollTop && scroll.y && target !== ref_headTable) {
        const scrollTop = target.scrollTop;
        if (ref_fixedColumnsBodyLeft && target !== ref_fixedColumnsBodyLeft) {
          ref_fixedColumnsBodyLeft.scrollTop = scrollTop;
        }
        if (ref_fixedColumnsBodyRight && target !== ref_fixedColumnsBodyRight) {
          ref_fixedColumnsBodyRight.scrollTop = scrollTop;
        }
        if (ref_bodyTable && target !== ref_bodyTable) {
          ref_bodyTable.scrollTop = scrollTop;
        }
      }
      // Remember last scrollTop for scroll direction detecting.
      this.lastScrollTop = target.scrollTop;
    },

    handleBodyScroll(e) {
      this.onScroll(e.target);
      this.handleBodyScrollLeft(e);
      this.handleBodyScrollTop(e);
    },
    handleWheel(event) {
      const { scroll = {} } = this.$props;
      if (window.navigator.userAgent.match(/Trident\/7\./) && scroll.y) {
        event.preventDefault();
        const wd = event.deltaY;
        const target = event.target;
        const {
          ref_bodyTable: bodyTable,
          ref_fixedColumnsBodyLeft: fixedColumnsBodyLeft,
          ref_fixedColumnsBodyRight: fixedColumnsBodyRight,
        } = this;
        let scrollTop = 0;

        if (this.lastScrollTop) {
          scrollTop = this.lastScrollTop + wd;
        } else {
          scrollTop = wd;
        }

        if (fixedColumnsBodyLeft && target !== fixedColumnsBodyLeft) {
          fixedColumnsBodyLeft.scrollTop = scrollTop;
        }
        if (fixedColumnsBodyRight && target !== fixedColumnsBodyRight) {
          fixedColumnsBodyRight.scrollTop = scrollTop;
        }
        if (bodyTable && target !== bodyTable) {
          bodyTable.scrollTop = scrollTop;
        }
      }
    },
    // saveChildrenRef(name, node) {
    //   this[`ref_${name}`] = node;
    // },
    saveRef(name) {
      return node => {
        this[`ref_${name}`] = node;
      };
    },

    saveTableNodeRef(node) {
      this.tableNode = node;
    },
    renderMainTable() {
      const { scroll, prefixCls } = this;
      const isAnyColumnsFixed = this.columnManager.isAnyColumnsFixed();
      const scrollable = isAnyColumnsFixed || scroll.x || scroll.y;

      const table = [
        this.renderTable({
          columns: this.columnManager.groupedColumns(),
          isAnyColumnsFixed,
        }),
        this.renderEmptyText(),
        this.renderFooter(),
      ];

      return scrollable ? (
        <ResizeObserver onResize={this.onFullTableResize}>
          <div class={`${prefixCls}-scroll`}>{table}</div>
        </ResizeObserver>) : table;
    },

    renderLeftFixedTable() {
      const { prefixCls } = this;

      return (
        <div key="left" class={`${prefixCls}-fixed-left`}>
          {this.renderTable({
            columns: this.columnManager.leftColumns(),
            fixed: 'left',
          })}
        </div>
      );
    },
    renderRightFixedTable() {
      const { prefixCls } = this;

      return (
        <div class={`${prefixCls}-fixed-right`}>
          {this.renderTable({
            columns: this.columnManager.rightColumns(),
            fixed: 'right',
          })}
        </div>
      );
    },

    renderTable(options) {
      const { columns, fixed, isAnyColumnsFixed } = options;
      const { prefixCls, scroll = {} } = this;
      const tableClassName = scroll.x || fixed ? `${prefixCls}-fixed` : '';

      const headTable = (
        <HeadTable
          key="head"
          columns={columns}
          fixed={fixed}
          tableClassName={tableClassName}
          handleBodyScrollLeft={this.handleBodyScrollLeft}
          expander={this.expander}
        />
      );

      const bodyTable = (
        <BodyTable
          key="body"
          columns={columns}
          fixed={fixed}
          tableClassName={tableClassName}
          getRowKey={this.getRowKey}
          handleWheel={this.handleWheel}
          handleBodyScroll={this.handleBodyScroll}
          expander={this.expander}
          isAnyColumnsFixed={isAnyColumnsFixed}
          ref="bodyRef"
        />
      );

      return [headTable, bodyTable];
    },

    renderTitle() {
      const { title, prefixCls, data } = this;
      return title ? (
        <div class={`${prefixCls}-title`} key="title">
          {title(data)}
        </div>
      ) : null;
    },

    renderFooter() {
      const { footer, prefixCls, data } = this;
      return footer ? (
        <div class={`${prefixCls}-footer`} key="footer">
          {footer(data)}
        </div>
      ) : null;
    },

    renderEmptyText() {
      const { emptyText, prefixCls, data } = this;
      if (data.length) {
        return null;
      }
      const emptyClassName = `${prefixCls}-placeholder`;
      return (
        <div class={emptyClassName} key="emptyText">
          {typeof emptyText === 'function' ? emptyText() : emptyText}
        </div>
      );
    },
    onColumnResize(columnKey, width){
        this.updateColsWidths(widths => {
          if (widths.get(columnKey) !== width) {
            const newWidths = new Map(widths);
            newWidths.set(columnKey, width);
            return newWidths;
          }
          return widths;
        });
    },
    onScroll(currentTarget){
      const { scrollWidth, clientWidth, scrollLeft } = currentTarget;
      this.pingedLeft = scrollLeft > 0;
      this.pingedRight = scrollLeft < scrollWidth - clientWidth;
    },
    onFullTableResize(){
      this.horizonScroll && this.onScroll(this.$refs.bodyRef.$el);
    },
  },

  render() {
    const props = getOptionProps(this);
    const { columnManager, getRowKey } = this;
    const prefixCls = props.prefixCls;

    const tableClassName = classNames(props.prefixCls, {
      [`${prefixCls}-fixed-header`]: props.useFixedHeader || (props.scroll && props.scroll.y),
      [`${prefixCls}-scroll-position-left ${prefixCls}-scroll-position-right`]:
        this.scrollPosition === 'both',
      [`${prefixCls}-scroll-position-${this.scrollPosition}`]: this.scrollPosition !== 'both',
      [`${prefixCls}-layout-fixed`]: this.isTableLayoutFixed(),
      [`${prefixCls}-ping-left`]: this.pingedLeft,
      [`${prefixCls}-ping-right`]: this.pingedRight,
    });

    const expandableTableProps = {
      props: {
        ...props,
        columnManager,
        getRowKey,
      },
      on: getListeners(this),
      scopedSlots: {
        default: expander => {
          this.expander = expander;
          return (
            <div
              {...{
                directives: [
                  {
                    name: 'ant-ref',
                    value: this.saveTableNodeRef,
                  },
                ],
              }}
              class={tableClassName}
              // style={props.style}
              // id={props.id}
            >
              {this.renderTitle()}
              <div class={`${prefixCls}-content`}>{this.renderMainTable()}</div>
            </div>
          );
        },
      },
    };
    return <ExpandableTable {...expandableTableProps} />;
  },
};
