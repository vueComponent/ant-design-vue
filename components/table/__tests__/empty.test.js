import { mount } from '@vue/test-utils';
import Table from '..';
import { nextTick } from 'vue';

const columns = [
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
];

const columnsFixed = [
  {
    title: 'Full Name',
    width: 100,
    dataIndex: 'name',
    key: 'name',
    fixed: 'left',
  },
  {
    title: 'Age',
    width: 100,
    dataIndex: 'age',
    key: 'age',
    fixed: 'left',
  },
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
  {
    title: 'Action',
    key: 'address',
    fixed: 'right',
    width: 100,
  },
];

describe('Table', () => {
  it('renders empty table', done => {
    const wrapper = mount(
      {
        render() {
          return <Table dataSource={[]} columns={columns} pagination={false} />;
        },
      },
      { sync: false },
    );
    nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });

  it('renders empty table with fixed columns', done => {
    const wrapper = mount(
      {
        render() {
          return <Table dataSource={[]} columns={columnsFixed} pagination={false} />;
        },
      },
      { sync: false },
    );
    nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });

  it('renders empty table with custom emptyText', done => {
    const wrapper = mount(
      {
        render() {
          return (
            <Table
              dataSource={[]}
              columns={columns}
              pagination={false}
              locale={{ emptyText: 'custom empty text ' }}
            />
          );
        },
      },
      { sync: false },
    );
    nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });

  it('renders empty table without emptyText when loading', done => {
    const wrapper = mount(
      {
        render() {
          return <Table dataSource={[]} columns={columns} loading />;
        },
      },
      { sync: false },
    );
    nextTick(() => {
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });
});
