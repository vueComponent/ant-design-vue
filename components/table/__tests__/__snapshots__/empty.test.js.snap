// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders empty table 1`] = `
<div class="ant-table-wrapper">
  <div class="ant-spin-nested-loading">
    <!---->
    <div class="ant-spin-container">
      <div class="ant-table ant-table-empty">
        <!---->
        <div class="ant-table-container">
          <div class="ant-table-content">
            <table style="table-layout: auto;">
              <colgroup></colgroup>
              <thead class="ant-table-thead">
                <tr>
                  <th class="ant-table-cell" colstart="0" colend="0">
                    <!---->Column 1
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="1" colend="1">
                    <!---->Column 2
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="2" colend="2">
                    <!---->Column 3
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="3" colend="3">
                    <!---->Column 4
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="4" colend="4">
                    <!---->Column 5
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="5" colend="5">
                    <!---->Column 6
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="6" colend="6">
                    <!---->Column 7
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="7" colend="7">
                    <!---->Column 8
                    <!---->
                  </th>
                </tr>
              </thead>
              <tbody class="ant-table-tbody">
                <!---->
                <tr class="ant-table-placeholder">
                  <td colspan="8" class="ant-table-cell">
                    <!---->
                    <div class="ant-empty ant-empty-normal">
                      <div class="ant-empty-image"><svg class="ant-empty-img-simple" width="64" height="41" viewBox="0 0 64 41">
                          <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
                            <ellipse class="ant-empty-img-simple-ellipse" fill="#F5F5F5" cx="32" cy="33" rx="32" ry="7"></ellipse>
                            <g class="ant-empty-img-simple-g" fill-rule="nonzero" stroke="#D9D9D9">
                              <path d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"></path>
                              <path d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z" fill="#FAFAFA" class="ant-empty-img-simple-path"></path>
                            </g>
                          </g>
                        </svg></div>
                      <p class="ant-empty-description">No Data</p>
                      <!---->
                    </div>
                    <!---->
                  </td>
                </tr>
              </tbody>
              <!---->
            </table>
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table with custom emptyText 1`] = `
<div class="ant-table-wrapper">
  <div class="ant-spin-nested-loading">
    <!---->
    <div class="ant-spin-container">
      <div class="ant-table ant-table-empty">
        <!---->
        <div class="ant-table-container">
          <div class="ant-table-content">
            <table style="table-layout: auto;">
              <colgroup></colgroup>
              <thead class="ant-table-thead">
                <tr>
                  <th class="ant-table-cell" colstart="0" colend="0">
                    <!---->Column 1
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="1" colend="1">
                    <!---->Column 2
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="2" colend="2">
                    <!---->Column 3
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="3" colend="3">
                    <!---->Column 4
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="4" colend="4">
                    <!---->Column 5
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="5" colend="5">
                    <!---->Column 6
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="6" colend="6">
                    <!---->Column 7
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="7" colend="7">
                    <!---->Column 8
                    <!---->
                  </th>
                </tr>
              </thead>
              <tbody class="ant-table-tbody">
                <!---->
                <tr class="ant-table-placeholder">
                  <td colspan="8" class="ant-table-cell">
                    <!---->custom empty text
                    <!---->
                  </td>
                </tr>
              </tbody>
              <!---->
            </table>
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table with fixed columns 1`] = `
<div class="ant-table-wrapper">
  <div class="ant-spin-nested-loading">
    <!---->
    <div class="ant-spin-container">
      <div class="ant-table ant-table-has-fix-left ant-table-has-fix-right ant-table-empty">
        <!---->
        <div class="ant-table-container">
          <div class="ant-table-content">
            <table style="table-layout: auto;">
              <colgroup>
                <col style="width: 100px;">
                <col style="width: 100px;">
                <col>
                <col>
                <col>
                <col>
                <col>
                <col>
                <col>
                <col>
                <col style="width: 100px;">
              </colgroup>
              <thead class="ant-table-thead">
                <tr>
                  <th class="ant-table-cell ant-table-cell-fix-left" style="position: sticky; left: 0px;" colstart="0" colend="0">
                    <!---->Full Name
                    <!---->
                  </th>
                  <th class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last" style="position: sticky; left: 0px;" colstart="1" colend="1">
                    <!---->Age
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="2" colend="2">
                    <!---->Column 1
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="3" colend="3">
                    <!---->Column 2
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="4" colend="4">
                    <!---->Column 3
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="5" colend="5">
                    <!---->Column 4
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="6" colend="6">
                    <!---->Column 5
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="7" colend="7">
                    <!---->Column 6
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="8" colend="8">
                    <!---->Column 7
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="9" colend="9">
                    <!---->Column 8
                    <!---->
                  </th>
                  <th class="ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first" style="position: sticky; right: 0px;" colstart="10" colend="10">
                    <!---->Action
                    <!---->
                  </th>
                </tr>
              </thead>
              <tbody class="ant-table-tbody">
                <!---->
                <tr class="ant-table-placeholder">
                  <td colspan="11" class="ant-table-cell">
                    <!---->
                    <div class="ant-empty ant-empty-normal">
                      <div class="ant-empty-image"><svg class="ant-empty-img-simple" width="64" height="41" viewBox="0 0 64 41">
                          <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
                            <ellipse class="ant-empty-img-simple-ellipse" fill="#F5F5F5" cx="32" cy="33" rx="32" ry="7"></ellipse>
                            <g class="ant-empty-img-simple-g" fill-rule="nonzero" stroke="#D9D9D9">
                              <path d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"></path>
                              <path d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z" fill="#FAFAFA" class="ant-empty-img-simple-path"></path>
                            </g>
                          </g>
                        </svg></div>
                      <p class="ant-empty-description">No Data</p>
                      <!---->
                    </div>
                    <!---->
                  </td>
                </tr>
              </tbody>
              <!---->
            </table>
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table without emptyText when loading 1`] = `
<div class="ant-table-wrapper">
  <div class="ant-spin-nested-loading">
    <div>
      <div class="ant-spin ant-spin-spinning" aria-live="polite" aria-busy="true"><span class="ant-spin-dot ant-spin-dot-spin"><i class="ant-spin-dot-item"></i><i class="ant-spin-dot-item"></i><i class="ant-spin-dot-item"></i><i class="ant-spin-dot-item"></i></span>
        <!---->
      </div>
    </div>
    <div class="ant-spin-container ant-spin-blur">
      <div class="ant-table ant-table-empty">
        <!---->
        <div class="ant-table-container">
          <div class="ant-table-content">
            <table style="table-layout: auto;">
              <colgroup></colgroup>
              <thead class="ant-table-thead">
                <tr>
                  <th class="ant-table-cell" colstart="0" colend="0">
                    <!---->Column 1
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="1" colend="1">
                    <!---->Column 2
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="2" colend="2">
                    <!---->Column 3
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="3" colend="3">
                    <!---->Column 4
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="4" colend="4">
                    <!---->Column 5
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="5" colend="5">
                    <!---->Column 6
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="6" colend="6">
                    <!---->Column 7
                    <!---->
                  </th>
                  <th class="ant-table-cell" colstart="7" colend="7">
                    <!---->Column 8
                    <!---->
                  </th>
                </tr>
              </thead>
              <tbody class="ant-table-tbody">
                <!---->
                <tr class="ant-table-placeholder">
                  <td colspan="8" class="ant-table-cell">
                    <!---->
                    <div class="ant-empty ant-empty-normal">
                      <div class="ant-empty-image"><svg class="ant-empty-img-simple" width="64" height="41" viewBox="0 0 64 41">
                          <g transform="translate(0 1)" fill="none" fill-rule="evenodd">
                            <ellipse class="ant-empty-img-simple-ellipse" fill="#F5F5F5" cx="32" cy="33" rx="32" ry="7"></ellipse>
                            <g class="ant-empty-img-simple-g" fill-rule="nonzero" stroke="#D9D9D9">
                              <path d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"></path>
                              <path d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z" fill="#FAFAFA" class="ant-empty-img-simple-path"></path>
                            </g>
                          </g>
                        </svg></div>
                      <p class="ant-empty-description">No Data</p>
                      <!---->
                    </div>
                    <!---->
                  </td>
                </tr>
              </tbody>
              <!---->
            </table>
          </div>
        </div>
        <!---->
      </div>
    </div>
  </div>
</div>
`;
